### PR TITLE
config with terraform

### PR DESCRIPTION
**Pull request by RahulMR42**
# Description

Manage policies and dynamic group and more using OCI Devops and Terraform.
Demonstrate how to execute terraform scripts from OCI Build runner.
Manage Terraform state management using object storage.

Fixes # (issues)

New samples # (demos / Illustrations )

Centralized configuration management using terraform (policies and dgs)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

- [x] Free Tier
- [x] Always Free
- [x] Paid 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings.
- [x] All the sensitive informations are masked or removed.
- [x] All the links mentioned are valid and tested.
- [x] Image names are separated only with '-'.
- [x] Added instruction to clone only this sample and links to other common sample pages.

### CODE REVIEW

**Comment by RahulMR42** [comment]
recheck

File: oci-config-examples/oci_devops_policy_dg_terraform/README.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+Allow dynamic-group "<DG Name>" to manage devops-family in c"<Compartment Name>"    
+Allow dynamic-group "<DG Name>" to manage generic-artifacts in c"<Compartment Name>"    
+Allow dynamic-group "<DG Name>" to use ons-topics in c"<Compartment Name>"  
+Allow dynamic-group "<DG Name>" to read secret-family in c"<Compartment Name>"
+```
+
+- Create a devops project - https://docs.oracle.com/en-us/iaas/Content/devops/using/create_project.htm#create_a_project 
+- Associate it with the notification topic created.
+
+![](images/oci_project.png)
```
Comment (ashok-cm) [inline oci-config-examples/oci_devops_policy_dg_terraform/README.md:59]: we may mask customer's name.

File: oci-config-examples/oci_devops_policy_dg_terraform/terraform/tenancy_scope/remote-backend_template.tf [hunk 2]
```diff
@@ -0,0 +1,13 @@
+terraform {
+  backend "s3" {
+    bucket   = "mr-prod-objectstore-for-terraform"
+    key      = "root_${compartment_name}/terraform.tfstate"
+    region   = "us-ashburn-1"
+    endpoint = "https://fahdabidiroottenancy.compat.objectstorage.us-ashburn-1.oraclecloud.com"
```
Comment (ashok-cm) [inline oci-config-examples/oci_devops_policy_dg_terraform/terraform/tenancy_scope/remote-backend_template.tf:6]: shall we mask the URL?

**Review by ashok-cm — CHANGES_REQUESTED** [review]

File: oci-config-examples/oci_devops_policy_dg_terraform/README.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+```
+ALL {resource.type = 'devopsbuildpipeline', resource.compartment.id = 'ocid1.compartment.oc1..xxx'} 
+ALL {resource.type = 'devopsrepository', resource.compartment.id = 'ocid1.compartment.oc1..xxx'}
+
+```
+
+- Create a policy with below statements.
+
+```
+Allow dynamic-group "<DG Name>" to manage repos in c"<Compartment Name>"
```
[the remark this answers is not in the document]
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/README.md:48]: haa ,its gone over a copy ,fixed .thank you

File: oci-config-examples/oci_devops_policy_dg_terraform/README.md (hunk 1 above)
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/README.md:59]: Done ,thank you

File: oci-config-examples/oci_devops_policy_dg_terraform/README.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+$ cd ../tenancy_scope 
+$ Edit variables.tf and update tenancy_ocid and region values.
+$ Edit remote-backend_template.tf and update the endpoint.
+```
+
+- Push the change back to OCI Code repo using using GIT CLI or UI over https or ssh authorization towards OCI Code repo  - https://docs.oracle.com/en-us/iaas/Content/devops/using/clone_repo.htm 
+
+![](images/oci_coderepo.png)
+
+- Let us do a build manual invokation.
```
[the remark this answers is not in the document]
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/README.md:148]: Done ,thank you

File: oci-config-examples/oci_devops_policy_dg_terraform/build_spec.yaml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+env:
+  # these are local variables to the build config
+  variables:
+    region: "us-ashburn-1" # Set your OCI Region.
+
+  # the value of a vaultVariable is the secret-id (in OCI ID format) stored in the OCI Vault service
+  # you can then access the value of that secret in your build_spec.yaml commands
+  vaultVariables:
+    aws_access_key_id: ocid1.vaultsecret.oc1.iad.amaaaaaak56z2vqaqsdelia6iiagjmtuok43w7ipklqwahwbwqvye52n74va
+    aws_secret_access_key: ocid1.vaultsecret.oc1.iad.amaaaaaak56z2vqa6vmu7h3ud3thzm7emsdb7fptb7zqmise2qlhx7xtehza
```
[the remark this answers is not in the document]
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/build_spec.yaml:15]: Done ,thank you

File: oci-config-examples/oci_devops_policy_dg_terraform/compartments.config [hunk 6]
```diff
@@ -0,0 +1,2 @@
+mr-prod-compartment=ocid1.compartment.oc1..aaaaaaaaexvr2lxjc5qjf7sfjemsyb7tbsgei3iplysjffslpnkqypx2sbmq
+mr-devops-canary=ocid1.compartment.oc1..aaaaaaaaievlqkktpe5yumlanr64gnzgi2vokbhrsuz2sddcjooewbqqj5ha
```
[the remark this answers is not in the document]
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/compartments.config:2]: Done ,thank you

File: oci-config-examples/oci_devops_policy_dg_terraform/terraform/compartment_scope/remote-backend_template.tf [hunk 7]
```diff
@@ -0,0 +1,13 @@
+terraform {
+  backend "s3" {
+    bucket   = "mr-prod-objectstore-for-terraform"
+    key      = "${compartment_name}/terraform.tfstate"
+    region   = "us-ashburn-1"
+    endpoint = "https://fahdabidiroottenancy.compat.objectstorage.us-ashburn-1.oraclecloud.com"
```
[the remark this answers is not in the document]
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/terraform/compartment_scope/remote-backend_template.tf:6]: Done ,thank you

File: oci-config-examples/oci_devops_policy_dg_terraform/terraform/compartment_scope/variables.tf [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+variable region{
+    type = string
+    default = "us-ashburn-1"
+}
+
+variable tenancy_ocid{
+    type = string
+    default = "ocid1.tenancy.oc1..aaaaaaaasu7rvefmsyk5kqczfmdqi5clpddejfjk2attdqnk6sbk72wajq5q"
```
[the remark this answers is not in the document]
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/terraform/compartment_scope/variables.tf:24]: Done ,thank you

File: oci-config-examples/oci_devops_policy_dg_terraform/terraform/tenancy_scope/variables.tf [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+
+variable region {
+    type = string
+    default = "us-ashburn-1"
+}
+
+variable tenancy_ocid{
+    type = string
+    default = "ocid1.tenancy.oc1..aaaaaaaasu7rvefmsyk5kqczfmdqi5clpddejfjk2attdqnk6sbk72wajq5q"
```
[the remark this answers is not in the document]
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/terraform/tenancy_scope/variables.tf:20]: Done ,thank you

File: oci-config-examples/oci_devops_policy_dg_terraform/README.md [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+![](images/oci_buildstage_2.png)
+
+- Create an object storage bucket - https://docs.oracle.com/en-us/iaas/Content/Object/home.htm
+
+![](images/oci_objectstore.png)
+
+- Create a customer secret keys for the user - https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/managingcredentials.htm#Working2
+
+![](images/oci_user_1.png)
```
[the remark this answers is not in the document]
Comment (RahulMR42) [reply oci-config-examples/oci_devops_policy_dg_terraform/README.md:85]: Done ,thank you

**Review by ashok-cm — APPROVED** [review]

**Pull request merged**